frontend/src/BenchmarkTest.res

```diff
@@ -163,73 +152,80 @@ let make = (
-      | None => "Unknown"
-      }
-
-      Belt.Map.Int.set(acc, index, tick)
-    })
+  let graph_metrics = React.useMemo1(() => {
+    dataByMetricName
+    ->Belt.Map.String.mapWithKey((metricName, (timeseries, metadata)) => {
+      let (comparisonTimeseries, comparisonMetadata) = Belt.Map.String.getWithDefault(
+        comparison,
+        metricName,
+        ([], []),
+      )
 
-    let annotations = if Belt.Array.length(comparisonTimeseries) > 0 {
-      let firstPullX = Belt.Array.length(comparisonTimeseries)
-      Some([
-        {
-          "series": "value",
-          "x": firstPullX,
-          "icon": "/branch.svg",
-          "text": "Open PR on GitHub",
-          "width": 21,
-          "height": 21,
-          "clickHandler": (_annotation, _point, _dygraph, _event) => {
-            switch pullNumber {
-            | Some(pullNumber) =>
-              DomHelpers.window->DomHelpers.windowOpen(
-                "https://github.com/" ++ repoId ++ "/pull/" ++ string_of_int(pullNumber),
-              )
-            | None => ()
-            }
+      let timeseries = Belt.Array.concat(comparisonTimeseries, timeseries)
+      let metadata = Belt.Array.concat(comparisonMetadata, metadata)
+
+      let xTicks = Belt.Array.reduceWithIndex(timeseries, Belt.Map.Int.empty, (acc, row, index) => {
+        // Use indexed instead of dates. This allows us to map to commits.
+        Belt.Array.set(row, 0, float_of_int(index))->ignore
+        let tick = switch Belt.Array.get(metadata, index) {
+        | Some(xMetadata) =>
+          let xValue = xMetadata["commit"]
+          DataHelpers.trimCommit(xValue)
+        | None => "Unknown"
+        }
+        Belt.Map.Int.set(acc, index, tick)
+      })
+
+      let annotations = if Belt.Array.length(comparisonTimeseries) > 0 {
+        let firstPullX = Belt.Array.length(comparisonTimeseries)
+        [
+          {
+            "series": "value",
+            "x": firstPullX,
+            "icon": "/branch.svg",
+            "text": "Open PR on GitHub",
+            "width": 21,
+            "height": 21,
+            "clickHandler": (_annotation, _point, _dygraph, _event) => {
+              switch pullNumber {
+              | Some(pullNumber) =>
+                DomHelpers.window->DomHelpers.windowOpen(
+                  "https://github.com/" ++ repoId ++ "/pull/" ++ string_of_int(pullNumber),
+                )
+              | None => ()
+              }
+            },
           },
-        },
-      ])
-    } else {
-      None
-    }
-
-    <LineGraph
-      onXLabelClick=goToCommitLink
-      onRender=onGraphRender
-      key=metricName
-      title=metricName
-      xTicks
-      data={timeseries->Belt.Array.sliceToEnd(-20)}
-      ?annotations
-      labels=["idx", "value"]
-    />
-  })
-  ->Belt.Map.String.valuesToArray
-  ->Rx.array
+        ]
+      } else {
+        []
+      }
+      (metricName, xTicks, timeseries->Belt.Array.sliceToEnd(-20), annotations)
```

Comment:
Should I use a record instead?

Reply:
Personally I don't see much point in (1) repackaging the data just to unpack it below and (2) iterating 2N times on the metrics instead of just N.

This is not very important of course, but I'd keep the `LineGraph` rendering here. Or do you find this clearer this way?

Reply:
I probably changed that when doing some explorations to optimize performances. But I agree that it doesn't make sense.